File: modules/anomaly_detector.py
"""
Anomaly Detection Module
Implements multiple anomaly detection algorithms for stock data
"""

import pandas as pd
import numpy as np
from sklearn.ensemble import IsolationForest
from sklearn.preprocessing import StandardScaler
from scipy import stats
from typing import List, Dict, Tuple


class AnomalyDetector:
    """Detects anomalies in stock data using multiple methods"""

    def __init__(self, method: str = 'z_score', **params):
        """
        Initialize anomaly detector

        Args:
            method: Detection method ('z_score', 'iqr', 'isolation_forest', 'bollinger', 'combined')
            **params: Method-specific parameters
        """
        self.method = method
        self.params = params
        self.anomalies = None

    def detect_anomalies(self, df: pd.DataFrame, features: List[str]) -> pd.DataFrame:
        """
        Detect anomalies in stock data

        Args:
            df: DataFrame with stock data
            features: List of features to analyze for anomalies

        Returns:
            DataFrame with anomaly flags and scores
        """
        df = df.copy()

        if self.method == 'z_score':
            df = self._z_score_detection(df, features)
        elif self.method == 'iqr':
            df = self._iqr_detection(df, features)
        elif self.method == 'isolation_forest':
            df = self._isolation_forest_detection(df, features)
        elif self.method == 'bollinger':
            df = self._bollinger_detection(df)
        elif self.method == 'combined':
            df = self._combined_detection(df, features)
        else:
            raise ValueError(f"Unknown method: {self.method}")

        return df

    def _z_score_detection(self, df: pd.DataFrame, features: List[str]) -> pd.DataFrame:
        """Detect anomalies using Z-score method"""
        threshold = self.params.get('z_threshold', 3.0)

        df['anomaly_score'] = 0
        df['is_anomaly'] = False
        df['anomaly_reasons'] = ''

        for feature in features:
            if feature not in df.columns:
                continue

            # Calculate Z-score
            mean = df[feature].mean()
            std = df[feature].std()

            if std == 0:
                continue

            z_scores = np.abs((df[feature] - mean) / std)
            feature_anomalies = z_scores > threshold

            # Update anomaly score (max z-score across features)
            df['anomaly_score'] = df['anomaly_score'].combine(z_scores, max)

            # Mark anomalies and add reasons
            mask = feature_anomalies
            df.loc[mask, 'is_anomaly'] = True
            df.loc[mask, 'anomaly_reasons'] += f"{feature} (Z={z_scores[mask].values[0]:.2f}); "

        return df

    def _iqr_detection(self, df: pd.DataFrame, features: List[str]) -> pd.DataFrame:
        """Detect anomalies using Interquartile Range (IQR) method"""
        multiplier = self.params.get('iqr_multiplier', 1.5)

        df['anomaly_score'] = 0
        df['is_anomaly'] = False
        df['anomaly_reasons'] = ''

        for feature in features:
            if feature not in df.columns:
                continue

            Q1 = df[feature].quantile(0.25)
            Q3 = df[feature].quantile(0.75)
            IQR = Q3 - Q1

            lower_bound = Q1 - multiplier * IQR
            upper_bound = Q3 + multiplier * IQR

            # Calculate how far outside bounds (normalized)
            below_lower = (lower_bound - df[feature]).clip(lower=0) / (IQR if IQR > 0 else 1)
            above_upper = (df[feature] - upper_bound).clip(lower=0) / (IQR if IQR > 0 else 1)
            deviation = below_lower + above_upper

            feature_anomalies = (df[feature] < lower_bound) | (df[feature] > upper_bound)

            # Update anomaly score
            df['anomaly_score'] = df['anomaly_score'].combine(deviation, max)

            # Mark anomalies and add reasons
            mask = feature_anomalies
            df.loc[mask, 'is_anomaly'] = True
            df.loc[mask, 'anomaly_reasons'] += f"{feature} (IQR outlier); "

        return df

    def _isolation_forest_detection(self, df: pd.DataFrame, features: List[str]) -> pd.DataFrame:
        """Detect anomalies using Isolation Forest"""
        contamination = self.params.get('contamination', 0.1)
        n_estimators = self.params.get('n_estimators', 100)

        # Prepare features
        valid_features = [f for f in features if f in df.columns]
        if not valid_features:
            df['anomaly_score'] = 0
            df['is_anomaly'] = False
            df['anomaly_reasons'] = ''
            return df

        # Remove NaN values for training
        df_clean = df[valid_features].dropna()

        if len(df_clean) < 10:
            df['anomaly_score'] = 0
            df['is_anomaly'] = False
            df['anomaly_reasons'] = ''
            return df

        # Standardize features
        scaler = StandardScaler()
        X_scaled = scaler.fit_transform(df_clean)

        # Train Isolation Forest
        iso_forest = IsolationForest(
            contamination=contamination,
            n_estimators=n_estimators,
            random_state=42
        )
        predictions = iso_forest.fit_predict(X_scaled)
        scores = iso_forest.score_samples(X_scaled)

        # Map back to original dataframe
        df['anomaly_score'] = 0.0
        df['is_anomaly'] = False
        df['anomaly_reasons'] = ''

        df.loc[df_clean.index, 'anomaly_score'] = -scores  # Invert so higher = more anomalous
        df.loc[df_clean.index, 'is_anomaly'] = predictions == -1
        df.loc[df.loc[df_clean.index, 'is_anomaly'].index, 'anomaly_reasons'] = 'Isolation Forest detection; '

        return df

    def _bollinger_detection(self, df: pd.DataFrame) -> pd.DataFrame:
        """Detect anomalies using Bollinger Bands"""
        df['anomaly_score'] = 0
        df['is_anomaly'] = False
        df['anomaly_reasons'] = ''

        if 'BB_Upper' in df.columns and 'BB_Lower' in df.columns:
            # Detect breaches
            upper_breach = df['Close'] > df['BB_Upper']
            lower_breach = df['Close'] < df['BB_Lower']

            # Calculate distance from bands (normalized)
            bb_range = df['BB_Upper'] - df['BB_Lower']
            upper_distance = ((df['Close'] - df['BB_Upper']) / bb_range).clip(lower=0)
            lower_distance = ((df['BB_Lower'] - df['Close']) / bb_range).clip(lower=0)

            df['anomaly_score'] = upper_distance + lower_distance

            df.loc[upper_breach, 'is_anomaly'] = True
            df.loc[upper_breach, 'anomaly_reasons'] += 'Above Bollinger Upper Band; '

            df.loc[lower_breach, 'is_anomaly'] = True
            df.loc[lower_breach, 'anomaly_reasons'] += 'Below Bollinger Lower Band; '

        return df

    def _combined_detection(self, df: pd.DataFrame, features: List[str]) -> pd.DataFrame:
        """Combine multiple detection methods"""
        voting_threshold = self.params.get('voting_threshold', 2)

        # Run multiple methods
        df_zscore = self._z_score_detection(df.copy(), features)
        df_iqr = self._iqr_detection(df.copy(), features)
        df_bollinger = self._bollinger_detection(df.copy())

        # Voting mechanism
        votes = (
            df_zscore['is_anomaly'].astype(int) +
            df_iqr['is_anomaly'].astype(int) +
            df_bollinger['is_anomaly'].astype(int)
        )

        df['is_anomaly'] = votes >= voting_threshold
        df['anomaly_score'] = (
            df_zscore['anomaly_score'] +
            df_iqr['anomaly_score'] +
            df_bollinger['anomaly_score']
        ) / 3

        # Combine reasons
        df['anomaly_reasons'] = ''
        for idx in df[df['is_anomaly']].index:
            reasons = []
            if df_zscore.loc[idx, 'is_anomaly']:
                reasons.append('Z-score')
            if df_iqr.loc[idx, 'is_anomaly']:
                reasons.append('IQR')
            if df_bollinger.loc[idx, 'is_anomaly']:
                reasons.append('Bollinger')
            df.loc[idx, 'anomaly_reasons'] = ', '.join(reasons)

        return df

    def get_anomaly_summary(self, df: pd.DataFrame) -> Dict:
        """Get summary statistics of detected anomalies"""
        anomalies = df[df['is_anomaly']]

        return {
            'total_anomalies': len(anomalies),
            'anomaly_rate': len(anomalies) / len(df) * 100 if len(df) > 0 else 0,
            'avg_anomaly_score': anomalies['anomaly_score'].mean() if len(anomalies) > 0 else 0,
            'max_anomaly_score': anomalies['anomaly_score'].max() if len(anomalies) > 0 else 0,
            'anomaly_dates': anomalies.index.tolist()
        }

    def get_top_anomalies(self, df: pd.DataFrame, n: int = 10) -> pd.DataFrame:
        """Get top N anomalies by score"""
        anomalies = df[df['is_anomaly']].copy()

        if len(anomalies) == 0:
            return pd.DataFrame()

        anomalies = anomalies.sort_values('anomaly_score', ascending=False)
        return anomalies.head(n)

File: lib/anomalyDetector.ts
/**
 * Anomaly Detection Module
 * Implements multiple anomaly detection algorithms
 */

import { StockDataPoint, AnomalyDataPoint, AnomalySummary, DetectionMethod } from './types';

export class AnomalyDetector {
  private method: DetectionMethod;
  private params: Record<string, number>;

  constructor(method: DetectionMethod, params: Record<string, number> = {}) {
    this.method = method;
    this.params = params;
  }

  /**
   * Detect anomalies in stock data
   */
  detectAnomalies(data: StockDataPoint[], features: string[]): AnomalyDataPoint[] {
    let result: AnomalyDataPoint[];

    switch (this.method) {
      case 'z_score':
        result = this.zScoreDetection(data, features);
        break;
      case 'iqr':
        result = this.iqrDetection(data, features);
        break;
      case 'isolation_forest':
        result = this.isolationForestDetection(data, features);
        break;
      case 'bollinger':
        result = this.bollingerDetection(data);
        break;
      case 'combined':
        result = this.combinedDetection(data, features);
        break;
      default:
        throw new Error(`Unknown method: ${this.method}`);
    }

    return result;
  }

  /**
   * Z-Score anomaly detection
   */
  private zScoreDetection(data: StockDataPoint[], features: string[]): AnomalyDataPoint[] {
    const threshold = this.params.zThreshold || 3.0;
    const result: AnomalyDataPoint[] = data.map((d) => ({
      ...d,
      anomalyScore: 0,
      isAnomaly: false,
      anomalyReasons: '',
    }));

    for (const feature of features) {
      const values = this.extractFeature(data, feature);
      if (values.length === 0) continue;

      const mean = values.reduce((a, b) => a + b, 0) / values.length;
      const std = Math.sqrt(
        values.reduce((a, b) => a + Math.pow(b - mean, 2), 0) / values.length
      );

      if (std === 0) continue;

      for (let i = 0; i < result.length; i++) {
        const value = this.getFeatureValue(result[i], feature);
        if (value === undefined) continue;

        const zScore = Math.abs((value - mean) / std);

        if (zScore > threshold) {
          result[i].isAnomaly = true;
          result[i].anomalyReasons += `${feature} (Z=${zScore.toFixed(2)}); `;
        }

        result[i].anomalyScore = Math.max(result[i].anomalyScore, zScore);
      }
    }

    return result;
  }

  /**
   * IQR anomaly detection
   */
  private iqrDetection(data: StockDataPoint[], features: string[]): AnomalyDataPoint[] {
    const multiplier = this.params.iqrMultiplier || 1.5;
    const result: AnomalyDataPoint[] = data.map((d) => ({
      ...d,
      anomalyScore: 0,
      isAnomaly: false,
      anomalyReasons: '',
    }));

    for (const feature of features) {
      const values = this.extractFeature(data, feature);
      if (values.length === 0) continue;

      const sorted = [...values].sort((a, b) => a - b);
      const q1Index = Math.floor(sorted.length * 0.25);
      const q3Index = Math.floor(sorted.length * 0.75);
      const q1 = sorted[q1Index];
      const q3 = sorted[q3Index];
      const iqr = q3 - q1;

      const lowerBound = q1 - multiplier * iqr;
      const upperBound = q3 + multiplier * iqr;

      for (let i = 0; i < result.length; i++) {
        const value = this.getFeatureValue(result[i], feature);
        if (value === undefined) continue;

        const belowLower = value < lowerBound ? (lowerBound - value) / (iqr || 1) : 0;
        const aboveUpper = value > upperBound ? (value - upperBound) / (iqr || 1) : 0;
        const deviation = belowLower + aboveUpper;

        if (value < lowerBound || value > upperBound) {
          result[i].isAnomaly = true;
          result[i].anomalyReasons += `${feature} (IQR outlier); `;
        }

        result[i].anomalyScore = Math.max(result[i].anomalyScore, deviation);
      }
    }

    return result;
  }

  /**
   * Simplified Isolation Forest detection
   * (Using a statistical approximation instead of full ML implementation)
   */
  private isolationForestDetection(data: StockDataPoint[], features: string[]): AnomalyDataPoint[] {
    const contamination = this.params.contamination || 0.1;
    const result: AnomalyDataPoint[] = data.map((d) => ({
      ...d,
      anomalyScore: 0,
      isAnomaly: false,
      anomalyReasons: '',
    }));

    // Extract feature matrix
    const featureMatrix: number[][] = [];
    for (let i = 0; i < data.length; i++) {
      const row: number[] = [];
      for (const feature of features) {
        const value = this.getFeatureValue(data[i], feature);
        if (value !== undefined) {
          row.push(value);
        }
      }
      if (row.length === features.length) {
        featureMatrix.push(row);
      } else {
        featureMatrix.push([]);
      }
    }

    // Normalize features
    const normalized = this.normalizeMatrix(featureMatrix);

    // Calculate anomaly scores based on distance from median
    for (let i = 0; i < normalized.length; i++) {
      if (normalized[i].length === 0) continue;

      const distances: number[] = [];
      for (let j = 0; j < normalized.length; j++) {
        if (i === j || normalized[j].length === 0) continue;
        const dist = this.euclideanDistance(normalized[i], normalized[j]);
        distances.push(dist);
      }

      if (distances.length === 0) continue;

      const avgDistance = distances.reduce((a, b) => a + b, 0) / distances.length;
      result[i].anomalyScore = avgDistance;
    }

    // Mark top contamination% as anomalies
    const scores = result.map((r, i) => ({ score: r.anomalyScore, index: i }));
    scores.sort((a, b) => b.score - a.score);
    const anomalyCount = Math.floor(data.length * contamination);

    for (let i = 0; i < anomalyCount; i++) {
      const index = scores[i].index;
      result[index].isAnomaly = true;
      result[index].anomalyReasons = 'Isolation Forest detection; ';
    }

    return result;
  }

  /**
   * Bollinger Bands anomaly detection
   */
  private bollingerDetection(data: StockDataPoint[]): AnomalyDataPoint[] {
    const result: AnomalyDataPoint[] = data.map((d) => ({
      ...d,
      anomalyScore: 0,
      isAnomaly: false,
      anomalyReasons: '',
    }));

    for (let i = 0; i < result.length; i++) {
      const point = result[i];

      if (point.bbUpper && point.bbLower) {
        const bbRange = point.bbUpper - point.bbLower;
        const upperDistance = point.close > point.bbUpper
          ? (point.close - point.bbUpper) / bbRange
          : 0;
        const lowerDistance = point.close < point.bbLower
          ? (point.bbLower - point.close) / bbRange
          : 0;

        point.anomalyScore = upperDistance + lowerDistance;

        if (point.close > point.bbUpper) {
          point.isAnomaly = true;
          point.anomalyReasons = 'Above Bollinger Upper Band; ';
        } else if (point.close < point.bbLower) {
          point.isAnomaly = true;
          point.anomalyReasons = 'Below Bollinger Lower Band; ';
        }
      }
    }

    return result;
  }

  /**
   * Combined detection using voting mechanism
   */
  private combinedDetection(data: StockDataPoint[], features: string[]): AnomalyDataPoint[] {
    const votingThreshold = this.params.votingThreshold || 2;

    // Run multiple methods
    const zScoreResults = this.zScoreDetection([...data], features);
    const iqrResults = this.iqrDetection([...data], features);
    const bollingerResults = this.bollingerDetection([...data]);

    const result: AnomalyDataPoint[] = data.map((d, i) => ({
      ...d,
      anomalyScore:
        (zScoreResults[i].anomalyScore +
         iqrResults[i].anomalyScore +
         bollingerResults[i].anomalyScore) / 3,
      isAnomaly: false,
      anomalyReasons: '',
    }));

    // Voting mechanism
    for (let i = 0; i < result.length; i++) {
      const votes =
        (zScoreResults[i].isAnomaly ? 1 : 0) +
        (iqrResults[i].isAnomaly ? 1 : 0) +
        (bollingerResults[i].isAnomaly ? 1 : 0);

      result[i].isAnomaly = votes >= votingThreshold;

      const methods: string[] = [];
      if (zScoreResults[i].isAnomaly) methods.push('Z-score');
      if (iqrResults[i].isAnomaly) methods.push('IQR');
      if (bollingerResults[i].isAnomaly) methods.push('Bollinger');

      result[i].anomalyReasons = methods.join(', ');
    }

    return result;
  }

  /**
   * Get anomaly summary statistics
   */
  getAnomalySummary(data: AnomalyDataPoint[]): AnomalySummary {
    const anomalies = data.filter((d) => d.isAnomaly);

    return {
      totalAnomalies: anomalies.length,
      anomalyRate: data.length > 0 ? (anomalies.length / data.length) * 100 : 0,
      avgAnomalyScore:
        anomalies.length > 0
          ? anomalies.reduce((sum, a) => sum + a.anomalyScore, 0) / anomalies.length
          : 0,
      maxAnomalyScore:
        anomalies.length > 0
          ? Math.max(...anomalies.map((a) => a.anomalyScore))
          : 0,
      anomalyDates: anomalies.map((a) => a.date),
    };
  }

  /**
   * Get top N anomalies by score
   */
  getTopAnomalies(data: AnomalyDataPoint[], n: number = 10): AnomalyDataPoint[] {
    const anomalies = data.filter((d) => d.isAnomaly);
    return anomalies
      .sort((a, b) => b.anomalyScore - a.anomalyScore)
      .slice(0, n);
  }

  // Helper methods

  private extractFeature(data: StockDataPoint[], feature: string): number[] {
    return data
      .map((d) => this.getFeatureValue(d, feature))
      .filter((v): v is number => v !== undefined);
  }

  private getFeatureValue(data: StockDataPoint, feature: string): number | undefined {
    const key = feature.toLowerCase();

    if (key === 'returns') return data.returns;
    if (key === 'volume_change' || key === 'volumechange') return data.volumeChange;
    if (key === 'range_pct' || key === 'rangepct') return data.rangePct;
    if (key === 'volatility') return data.volatility;

    return undefined;
  }

  private normalizeMatrix(matrix: number[][]): number[][] {
    if (matrix.length === 0) return [];

    const numFeatures = matrix[0].length;
    const normalized: number[][] = [];

    // Calculate min and max for each feature
    const mins: number[] = [];
    const maxs: number[] = [];

    for (let j = 0; j < numFeatures; j++) {
      const column = matrix.map((row) => row[j]).filter((v) => v !== undefined);
      mins.push(Math.min(...column));
      maxs.push(Math.max(...column));
    }

    // Normalize
    for (const row of matrix) {
      if (row.length === 0) {
        normalized.push([]);
        continue;
      }

      const normalizedRow: number[] = [];
      for (let j = 0; j < row.length; j++) {
        const range = maxs[j] - mins[j];
        if (range === 0) {
          normalizedRow.push(0);
        } else {
          normalizedRow.push((row[j] - mins[j]) / range);
        }
      }
      normalized.push(normalizedRow);
    }

    return normalized;
  }

  private euclideanDistance(a: number[], b: number[]): number {
    let sum = 0;
    for (let i = 0; i < a.length; i++) {
      sum += Math.pow(a[i] - b[i], 2);
    }
    return Math.sqrt(sum);
  }
}
